Refactor/unify arg popping

benlesh opened the pull request:
fix(from): objects that are thennable that happen to have a subscribe method will no longer error.

refactor: Unify argument parsing and runtime type checks
    
- Ensures that all runtime checks that check to see if a value is a function are done via `isFunction`.
- Unifies a common pattern of inspecting the last argument and popping it off as `popScheduler` and `popResultSelector`.


~ 2KB in savings

(Down to the ~52K range)

cartant commented on src/internal/scheduled/scheduleIterable.ts:
```diff
@@ -1,45 +1,49 @@
-      if (iterator && typeof iterator.return === 'function') {
-        iterator.return();
-      }
+      isFunction(iterator.return) && iterator.return();
```
issue: The replaced code tests `iterator`, so optional chaining should be used, IMO.
```
isFunction(iterator?.return)
```

cartant commented on src/internal/observable/interval.ts:
```diff
@@ -52,20 +54,17 @@ import { Subscriber } from '../Subscriber';
-    );
+  return new Observable<number>((subscriber) => {
+    subscriber.add(scheduler.schedule(dispatch as any, period, { subscriber, counter: 0, period }));
     return subscriber;
```
issue: Unrelated to the change, but I don't think it's necessary to return the subscriber here.

cartant commented on src/internal/observable/from.ts:
```diff
@@ -130,17 +131,16 @@ export function from<T>(input: ObservableInput<T>, scheduler?: SchedulerLike): O
 
 function subscribeTo<T>(result: ObservableInput<T>): (subscriber: Subscriber<T>) => Subscription | void {
-  if (result && typeof (result as any)[Symbol_observable] === 'function') {
+  if (isFunction((result as any)[Symbol_observable])) {
```
issue: This isn't quite the same. `result` is checked, so I think optional chaining is needed to effect exactly the same behaviour:
```
isFunction((result as any)?.[Symbol_observable])
```

benlesh replied:
I've opted instead to just check `result != null` at the start. As it would be more efficient.

cartant commented on src/internal/observable/from.ts:
```diff
@@ -130,17 +131,16 @@ export function from<T>(input: ObservableInput<T>, scheduler?: SchedulerLike): O
   } else if (isPromise(result)) {
     return subscribeToPromise(result);
-  } else if (result && typeof (result as any)[Symbol_iterator] === 'function') {
+  } else if (isFunction((result as any)[Symbol_iterator])) {
```
issue: Optional chaining again.

cartant reviewed (changes requested):
AFAICT, if this PR is to exactly reproduce the checks of the existing implementation, there are a bunch of places where optional chaining needs to be used.

benlesh commented on src/internal/scheduled/scheduled.ts:
```diff
@@ -29,11 +31,11 @@ export function scheduled<T>(input: ObservableInput<T>, scheduler: SchedulerLike
+    } else if (isIterable(input) || typeof input === 'string') {
       return scheduleIterable(input, scheduler);
-    } else if (Symbol && Symbol.asyncIterator && typeof (input as any)[Symbol.asyncIterator] === 'function') {
+    } else if (isFunction((input as any)[Symbol.asyncIterator])) {
```
Yeah, we can probably do that. IE is the last problem, and people supporting that should have about 5 years of experience adding polyfills.

benlesh commented:
A little further down a rabbit hole while addressing your comments: I've refactored `from` and `scheduled` quite a bit, and there's more to do in a follow-up PR as noted in some todos.

From the commit message:

refactor: `from` and `scheduled` use the same methodology
    
- Ensures that `from` and `scheduled` do checks in the same order. Note that I realize that scheduled has no check for `input instanceof Observable`, however, the interop check right after is about the same thing, and I'm not terribly worried about the performance here for the moment. Any optimizations in this area can come once we fully remove the deprecated `from(x, scheduler)` pathway, so we don't bloat that hot-path.
- Unifies the error thrown when a non-observable argument is passed to either `from` or `scheduled`.
- Unifies checks for `isAsyncIterable` and `isAsyncIterable`.
- Removes extra boilerplate around `subscribeTo` logic, as we don't use it inside of any other operators, it was just unnecessary overhead.
- Moves functions closer to actual usage.
- Adds a couple of TODOs around `innerFrom`. Since we're using `from` throughout the library now to convert our returned `ObservableInputs` for inner subscriptions, we are now going to inadvertently include `scheduled` in many operators. We need to switch those operators to use `innerFrom` where possible, as it will omit the bit `scheduled` logic, because it's unnecessary.

benlesh commented:
Rebased and resolved conflicts.

cartant commented on src/internal/observable/from.ts:
```diff
@@ -119,34 +119,130 @@ export function from<O extends ObservableInput<any>>(input: O, scheduler: Schedu
+    //    This is a known issue, but considered an edge case. The alternative would
+    //    be to copy the array before executing the loop, but this has
+    //    performance implications.
+    for (let i = 0; i < array.length && !subscriber.closed; i++) {
```
question: Would there be any advantage in using `for ... of` here? 

```ts
for (const value of array) {
  if (subscriber.closed) break;
  subscriber.next(value);
}
```

benlesh replied:
The primary reason is that this works for any ArrayLike. `for...of` will not.

Also, `for...of`, last I checked, was notably slower than the `for(;;)` loop, as it uses the iterator instead of accessing the array by index. However, there's an argument to be made for condensing the `fromIterable` and `fromArray` code into one thing, but again, that doesn't work for every ArrayLike. We'd have to do an `Array.from()` first, and then iterate it, which would cause double iteration and slow us down.

cartant reviewed (approved):
LGTM. Just a speculative question.

tonivj5 commented on src/internal/observable/from.ts:
```diff
@@ -119,34 +119,130 @@ export function from<O extends ObservableInput<any>>(input: O, scheduler: Schedu
+ */
+function fromInteropObservable<T>(obj: any) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    const obs = (obj as any)[Symbol_observable]();
+    if (typeof obs.subscribe !== 'function') {
+      // Should be caught by observable subscribe function error handling.
+      throw new TypeError('Provided object does not correctly implement Symbol.observable');
+    } else {
+      return obs.subscribe(subscriber);
+    }
```
```suggestion
    const obs = obj[Symbol_observable]();
    if (isFunction(obs.subscribe)) {
      return obs.subscribe(subscriber);
    }
    // Should be caught by observable subscribe function error handling.
    throw new TypeError('Provided object does not correctly implement Symbol.observable');
```

`obj` is `any` already, isn't it?

tonivj5 commented on src/internal/util/isObservable.ts:
```diff
@@ -1,9 +1,11 @@
  */
 export function isObservable<T>(obj: any): obj is Observable<T> {
-  return !!obj && (obj instanceof Observable || (typeof obj.lift === 'function' && typeof obj.subscribe === 'function'));
+  return !!obj && (obj instanceof Observable || (isFunction(obj.lift) && isFunction(obj.subscribe)));
```
```suggestion
  return obj && (obj instanceof Observable || (isFunction(obj.lift) && isFunction(obj.subscribe)));
```

benlesh replied:
So this change actually illustrates why I'm not always a fan of the inline comment changes. I looked and thought "yeah, that's reasonable" and merged it, and it broke a test :)

I've reverted that change and added a comment. But basically, this is a publicly exposed API, and we want to make sure it returns a boolean.... but if someone passes `null`, it would return `null` and not `false` after this change.

tonivj5 commented on src/internal/scheduled/scheduled.ts:
```diff
@@ -25,15 +28,19 @@ export function scheduled<T>(input: ObservableInput<T>, scheduler: SchedulerLike
-    } else if (Symbol && Symbol.asyncIterator && typeof (input as any)[Symbol.asyncIterator] === 'function') {
+    }
+    if (isAsyncIterable(input)) {
       return scheduleAsyncIterable(input as any, scheduler);
```
```suggestion
      return scheduleAsyncIterable(input, scheduler);
```

Using `isAsyncIterable` as type guard, `input` shouldn't be `AsyncIterable` already?

tonivj5 commented on src/internal/observable/from.ts:
```diff
@@ -119,34 +119,130 @@ export function from<O extends ObservableInput<any>>(input: O, scheduler: Schedu
+    if (isPromise(input)) {
+      return fromPromise(input);
+    }
+    if (isIterable(input)) {
+      return fromIterable(input as any);
+    }
+    if (isAsyncIterable(input)) {
+      return fromAsyncIterable(input as any);
```
Same `fromAsyncIterable` and `fromIterable`?

benlesh replied:
Yeah... although looking at those, I'm going to need to reverse their order I think. 🤔
